sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala

```diff
@@ -427,6 +455,22 @@ object GpuShuffleExchangeExecBase {
                 batch.close()
               }
             }
+
+            // Check if this is the first call and iterator has no more data
+            // If so, set the ThreadLocal flag indicating single batch task
+            if (isFirstCall) {
+              isFirstCall = false
+
+              if (!iter.hasNext) {
```

Comment:
we have to ensure that hasNext will not cause the whole plan to execute, every circumstance.

Reply:
> I think the biggest problem may lie in Scan part, typically we'll do real reading in hasNext().

Yes this is the type of issue I was expecting to find.  Though in the example you provided, the function returns an iterator (`CachedGpuBatchIterator`), which seems inert in the `hasNext` (just checks `pending.nonEmpty`, which was needed to instantiate the iterator object)

What I meant by checking every case is it has to be exhaustive, and can be pretty hard to get right. Sometimes we extend Iterator, sometimes we just do it inline: `new Iterator[T] { ... }`. I do not know if there are clever ways of finding all the issues here.